src/impl_/pyclass/lazy_type_object.rs

```diff
@@ -43,39 +44,33 @@ impl<T> LazyTypeObject<T> {
 
 impl<T: PyClass> LazyTypeObject<T> {
     /// Gets the type object contained by this `LazyTypeObject`, initializing it if needed.
-    pub fn get_or_init(&self, py: Python<'_>) -> *mut ffi::PyTypeObject {
+    pub fn get_or_init<'py>(&'py self, py: Python<'py>) -> &'py PyType {
         self.get_or_try_init(py).unwrap_or_else(|err| {
             err.print(py);
             panic!("failed to create type object for {}", T::NAME)
         })
     }
 
     /// Fallible version of the above.
-    pub(crate) fn get_or_try_init(&self, py: Python<'_>) -> PyResult<*mut ffi::PyTypeObject> {
-        fn inner<T: PyClass>() -> PyResult<*mut ffi::PyTypeObject> {
-            // Safety: `py` is held by the caller of `get_or_init`.
-            let py = unsafe { Python::assume_gil_acquired() };
-            create_type_object::<T>(py)
-        }
-
+    pub(crate) fn get_or_try_init<'py>(&'py self, py: Python<'py>) -> PyResult<&'py PyType> {
         self.0
-            .get_or_try_init(py, inner::<T>, T::NAME, T::items_iter())
+            .get_or_try_init(py, create_type_object::<T>, T::NAME, T::items_iter())
     }
 }
 
 impl LazyTypeObjectInner {
-    fn get_or_try_init(
-        &self,
-        py: Python<'_>,
-        init: fn() -> PyResult<*mut ffi::PyTypeObject>,
+    fn get_or_try_init<'py>(
+        &'py self,
+        py: Python<'py>,
+        init: fn(Python<'py>) -> PyResult<Py<PyType>>,
```

Comment:
Does this imply dynamic dispatch through a function pointer? Is that actually the right thing to do since this is only called during the initialization slow path?

Reply:
Yes, I think the cost of dynamic dispatch is unlikely to be relevant when it's only called once for each type? Or am I missing something here... ?

Reply:
No I don't think so. To be honest, this is mainly me always getting confused about function pointers and items (the singleton type of each function) and traits in Rust.